### ISSUE TITLE
softGlue fails to build for win32-x86 and windows-x64 against base-3.15.4

### ISSUE THREAD

**Issue by kmpeters** [epics-modules/softGlue#1]
I'm using the following:

```
Windows 7 (64-bit)
Visual Studio 2010 Professional & Visual Studio 2015 Community Edition
Make 4.1 (from the EPICS web site)
Strawberry perl 5.20.3001
```

The softGlue trunk (5761d73170389a340a53bfe9fe87ee2cfe6c2d52) fails to build for win32-x86 and windows-x64 with the following errors:

```
link -nologo /subsystem:windows /dll -LTCG -incremental:no -opt:ref -release       -out:softGlue.dll -implib:softGlue.lib         drvIP_EP201.obj devAsynSoftGlu
e.obj      D:/synApps-testing/VS10/3.15/synApps/support/asyn/lib/win32-x86/asyn.lib  D:/synApps-testing/VS10/3.15/synApps/support/ipac/lib/win32-x86/Ipac.lib  D
:/synApps-testing/VS10/3.15/base-3.15.4/lib/win32-x86/dbRecStd.lib  D:/synApps-testing/VS10/3.15/base-3.15.4/lib/win32-x86/dbCore.lib  D:/synApps-testing/VS10/3
.15/base-3.15.4/lib/win32-x86/ca.lib  D:/synApps-testing/VS10/3.15/base-3.15.4/lib/win32-x86/Com.lib
   Creating library softGlue.lib and object softGlue.exp
drvIP_EP201.obj : error LNK2001: unresolved external symbol _pasynUInt32DigitalBase
drvIP_EP201.obj : error LNK2001: unresolved external symbol _pasynManager
drvIP_EP201.obj : error LNK2001: unresolved external symbol _pasynTrace
drvIP_EP201.obj : error LNK2001: unresolved external symbol _pasynInt32Base
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol _pasynEpicsUtils
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol _pdbbase
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol _interruptAccept
softGlue.dll : fatal error LNK1120: 7 unresolved externals
D:/synApps-testing/VS10/3.15/base-3.15.4/configure/RULES_BUILD:281: recipe for target 'softGlue.dll' failed
make[4]: *** [softGlue.dll] Error 1120
make[4]: Leaving directory 'D:/synApps-testing/VS10/3.15/synApps/support/softGlue/softGlueApp/src/O.win32-x86'
D:/synApps-testing/VS10/3.15/base-3.15.4/configure/RULES_ARCHS:61: recipe for target 'install.win32-x86' failed
make[3]: *** [install.win32-x86] Error 2
make[3]: Leaving directory 'D:/synApps-testing/VS10/3.15/synApps/support/softGlue/softGlueApp/src'
D:/synApps-testing/VS10/3.15/base-3.15.4/configure/RULES_DIRS:88: recipe for target 'src.install' failed
make[2]: *** [src.install] Error 2
make[2]: Leaving directory 'D:/synApps-testing/VS10/3.15/synApps/support/softGlue/softGlueApp'
D:/synApps-testing/VS10/3.15/base-3.15.4/configure/RULES_DIRS:88: recipe for target 'softGlueApp.install' failed
make[1]: *** [softGlueApp.install] Error 2
make[1]: Leaving directory 'D:/synApps-testing/VS10/3.15/synApps/support/softGlue'
D:/synApps-testing/VS10/3.15/base-3.15.4/configure/RULES_DIRS:88: recipe for target 'D:/synApps-testing/VS10/3.15/synApps/support/softGlue.install' failed
make: *** [D:/synApps-testing/VS10/3.15/synApps/support/softGlue.install] Error 2
```

Builds for win32-x86-static and windows-x64-static succeed.  Builds against base 3.14.12.5 succeed for all Windows arches.

This problem looks similar to https://github.com/epics-modules/sscan/issues/3, but epicsShareAPI doesn't appear in softGlue.

**Comment by anjohnson** [comment]
Any time you have a Windows build that works on the `-static` target but fails to link on the dynamic target like this, the issue is almost invariably related to name decoration.

If you look at the unresolved symbol name from epics-modules/sscan#3 the name it was looking for was `__imp__dbNameToAddr@8`. The `__imp__` part at the beginning says that this is an import from another DLL, and I suspect the `@8` at the end probably relates to the calling convention. In the errors here, the symbols it's looking for all start with just a single `_`. That and the absence of a leading `__imp__` tells me that here the problem is that those named functions are not properly decorated. They do have an `epicsShareFunc` keyword in their headers, but the line

``` c
#define epicsExportSharedSymbols
```

or

``` c
#include "epicsExport.h"
```

is in the wrong place compared to the asyn `#include` lines, so the decorator is giving the compiler the wrong information (`epicsShareFunc` and some of the other macros expand out to two different things depending on whether they're found before or after that `#define`). In fact it seems that the `drvIP_EP201.c` file doesn't seem to have either of the above lines, so I would fix that first (and reorganize the include lines, see below). Recent Microsoft compilers seem to be able to cope with a certain number of wrongly-decorated functions, but after so many it seems to give up and make you fix them in the source. If you build for Windows using gcc (under Cygwin, native -mingw or a Linux -mingw cross-build) you will be forced to fix them anyway).

There is a logical and relatively easy way to structure your include statements which ensures that this problem doesn't bite you, provided you stick to these rules:

``` c
#include <system headers>

#include <headers from other libraries>

#define epicsExportSharedSymbols // or #include "epicsExport.h"

#include <headers from this library>

// Code...
```

If any of the `<headers from this library>` also `#include <headers from other libraries>` and your code uses any symbols declared in those headers, you _must_ ensure that they are also included in the set of `<headers from other libraries>` included above the `epicsExportSharedSymbols` line.

You might see source files that save and `#undef` the `epicsExportSharedSymbols` macro and then redefine it again later; don't try and learn from them, that approach makes it hard to work out where you've made a mistake in your lists of headers.

**Comment by timmmooney** [comment]
There may be a race condition in our consideration of this problem.  I found that drvIP_EP201.c had "#include epicsExport.h" in the wrong place, and included drvIP_EP201.h, which also had it in the wrong place.  I deleted the directive from drvIP_EP201.c, and fixed its placement in drvIP_EP201.h.
But I didn't comment here yet, because I was waiting to see if this fixed Kevin's build.

**Comment by anjohnson** [comment]
@timmmooney I suspect that no header file should ever `#include <epicsExport.h>` anyway. In fact I now regret that we ever made that file do the `#define epicsExportSharedSymbols` thing because it is so easy to lose sight of the special nature of that particular header. The `#define` version stands out reasonably well, but the `#include` line normally looks like just another header file.

**Comment by MarkRivers** [comment]
Tim's change from earlier today does not fix the problem.  I'm working on this now, I have fixed quite a few of these problems in the past.

**Comment by MarkRivers** [comment]
I have made a new rivers_fix branch.  I changed the files so they look OK to me.  epicsExport.h is in the correct location, only definitions for items in this DLL come after that.

However, it still does not work, the symbols from asyn and base are not found.  Very similar looking modules like busy build OK.

link -nologo /subsystem:windows /dll -LTCG -incremental:no -opt:ref -release  /MACHINE:X64      -out:softGlue.dll -implib:softGlue.lib         drvIP_EP201.obj devAsynSoftGlue.obj
    J:/epics/devel-3.15/asyn-4-30/lib/windows-x64/asyn.lib  J:/epics/devel-3.15/ipac-2-13/lib/windows-x64/Ipac.lib  H:/epics/base-3.15.4/lib/windows-x64/dbRecStd.lib  H:/epics/base
-3.15.4/lib/windows-x64/dbCore.lib  H:/epics/base-3.15.4/lib/windows-x64/ca.lib  H:/epics/base-3.15.4/lib/windows-x64/Com.lib
   Creating library softGlue.lib and object softGlue.exp
drvIP_EP201.obj : error LNK2001: unresolved external symbol pasynUInt32DigitalBase
drvIP_EP201.obj : error LNK2001: unresolved external symbol pasynManager
drvIP_EP201.obj : error LNK2001: unresolved external symbol pasynTrace
drvIP_EP201.obj : error LNK2001: unresolved external symbol pasynInt32Base
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol pasynEpicsUtils
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol pdbbase
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol interruptAccept
softGlue.dll : fatal error LNK1120: 7 unresolved externals
H:/epics/base-3.15.4/configure/RULES_BUILD:281: recipe for target 'softGlue.dll' failed
make[1]: **\* [softGlue.dll] Error 1120
make[1]: Leaving directory 'J:/epics/devel-3.15/softGlue/softGlueApp/src/O.windows-x64'
H:/epics/base-3.15.4/configure/RULES_ARCHS:61: recipe for target 'install.windows-x64' failed
make: **\* [install.windows-x64] Error 2

@anjohnson I have a questions about files like boRecord.h.  They #include <epicsExport.h>.  By having them in the list of include files why doesn't this screw things up, because they will define epicsExportSharedSymbols?  It does seem to work OK, I just don't understand why.

@anjohnson can you look at my branch and see what is wrong?  I am stumped.

**Comment by MarkRivers** [comment]
This problem appears to be specific to 3.15.4.  I just built the softGlue master branch and rivers_fix branch with no problems on 3.14.12.5.

**3.14.12.5 rivers_fix branch:**

J:\epics\devel\softGlue\softGlueApp\src>make
make -C O.win32-x86 -f ../Makefile TOP=../../.. \
    T_A=win32-x86 install
make[1]: Entering directory 'J:/epics/devel/softGlue/softGlueApp/src/O.win32-x86'
cl -c                /nologo /D__STDC__=0 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_NONSTDC_NO_DEPRECATE   /Ox /GL   /W3 /w44355
        /MT     -I. -I..\O.Common -I. -I.. -I..\..\..\include\os\WIN32 -I..\..\..\include    -IJ:\epics\devel
\asyn-4-30\include  -IJ:\epics\devel\ipac-2-13\include   -IH:\epics\base-3.14.12.5\include\os\WIN32 -IH:\epi
cs\base-3.14.12.5\include        ..\drvIP_EP201.c
drvIP_EP201.c
cl -c                /nologo /D__STDC__=0 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_NONSTDC_NO_DEPRECATE   /Ox /GL   /W3 /w44355
        /MT     -I. -I..\O.Common -I. -I.. -I..\..\..\include\os\WIN32 -I..\..\..\include    -IJ:\epics\devel
\asyn-4-30\include  -IJ:\epics\devel\ipac-2-13\include   -IH:\epics\base-3.14.12.5\include\os\WIN32 -IH:\epi
cs\base-3.14.12.5\include        ..\devAsynSoftGlue.c
devAsynSoftGlue.c
link /nologo /subsystem:windows /dll /LTCG /incremental:no /opt:ref /release        /implib:softGlue.lib /out:softGlue.d
ll          drvIP_EP201.obj devAsynSoftGlue.obj      J:\epics\devel\asyn-4-30\lib\win32-x86\asyn.lib  J:\epics\d
evel\ipac-2-13\lib\win32-x86\Ipac.lib  H:\epics\base-3.14.12.5\lib\win32-x86\recIoc.lib  H:\epics\base-3.14.1
2.5\lib\win32-x86\softDevIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\miscIoc.lib  H:\epics\base-3.14.12.5\
lib\win32-x86\rsrvIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\dbtoolsIoc.lib  H:\epics\base-3.14.12.5\lib\
win32-x86\asIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\dbIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\r
egistryIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\dbStaticIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\c
a.lib  H:\epics\base-3.14.12.5\lib\win32-x86\Com.lib
   Creating library softGlue.lib and object softGlue.exp
Generating code
Finished generating code
"Installing shared library ../../../bin/win32-x86/softGlue.dll"
mkdir ../../../bin
mkdir ../../../bin/win32-x86
"Installing library ../../../lib/win32-x86/softGlue.lib"
mkdir ../../../lib
mkdir ../../../lib/win32-x86
make[1]: Leaving directory 'J:/epics/devel/softGlue/softGlueApp/src/O.win32-x86'

**3.14.12.5 master branch:**

J:\epics\devel\softGlue\softGlueApp\src>make clean
perl -MExtUtils::Command -e rm_rf O.win32-x86 O.Common

J:\epics\devel\softGlue\softGlueApp\src>make
perl H:/epics/base-3.14.12.5/bin/win32-x86/makeMakefile.pl O.win32-x86 ../../..
perl -MExtUtils::Command -e mkpath O.Common
make -C O.win32-x86 -f ../Makefile TOP=../../.. \
    T_A=win32-x86 install
make[1]: Entering directory 'J:/epics/devel/softGlue/softGlueApp/src/O.win32-x86'
"Installing generic include file ../../../include/drvIP_EP201.h"
cl -c                /nologo /D__STDC__=0 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_NONSTDC_NO_DEPRECATE   /Ox /GL   /W3 /w44355
        /MT     -I. -I..\O.Common -I. -I.. -I..\..\..\include\os\WIN32 -I..\..\..\include    -IJ:\epics\devel
\asyn-4-30\include  -IJ:\epics\devel\ipac-2-13\include   -IH:\epics\base-3.14.12.5\include\os\WIN32 -IH:\epi
cs\base-3.14.12.5\include        ..\drvIP_EP201.c
drvIP_EP201.c
cl -c                /nologo /D__STDC__=0 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_NONSTDC_NO_DEPRECATE   /Ox /GL   /W3 /w44355
        /MT     -I. -I..\O.Common -I. -I.. -I..\..\..\include\os\WIN32 -I..\..\..\include    -IJ:\epics\devel
\asyn-4-30\include  -IJ:\epics\devel\ipac-2-13\include   -IH:\epics\base-3.14.12.5\include\os\WIN32 -IH:\epi
cs\base-3.14.12.5\include        ..\devAsynSoftGlue.c
devAsynSoftGlue.c
link /nologo /subsystem:windows /dll /LTCG /incremental:no /opt:ref /release        /implib:softGlue.lib /out:softGlue.d
ll          drvIP_EP201.obj devAsynSoftGlue.obj      J:\epics\devel\asyn-4-30\lib\win32-x86\asyn.lib  J:\epics\d
evel\ipac-2-13\lib\win32-x86\Ipac.lib  H:\epics\base-3.14.12.5\lib\win32-x86\recIoc.lib  H:\epics\base-3.14.1
2.5\lib\win32-x86\softDevIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\miscIoc.lib  H:\epics\base-3.14.12.5\
lib\win32-x86\rsrvIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\dbtoolsIoc.lib  H:\epics\base-3.14.12.5\lib\
win32-x86\asIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\dbIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\r
egistryIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\dbStaticIoc.lib  H:\epics\base-3.14.12.5\lib\win32-x86\c
a.lib  H:\epics\base-3.14.12.5\lib\win32-x86\Com.lib
   Creating library softGlue.lib and object softGlue.exp
Generating code
Finished generating code
"Installing shared library ../../../bin/win32-x86/softGlue.dll"
"Installing library ../../../lib/win32-x86/softGlue.lib"
make[1]: Leaving directory 'J:/epics/devel/softGlue/softGlueApp/src/O.win32-x86'

So it is either a difference in base or asyn (which is the same source) that it causing to fail in 3.15.4 and work in 3.14.12.5.

**3.15.4 rivers_fix branch**

J:\epics\devel-3.15\softGlue\softGlueApp\src>make clean
perl -CSD -MExtUtils::Command -e rm_rf O.win32-x86 O.Common

J:\epics\devel-3.15\softGlue\softGlueApp\src>make
perl -CSD H:/epics/base-3.15.4/bin/win32-x86/makeMakefile.pl O.win32-x86 ../../..
perl -CSD -MExtUtils::Command -e mkpath O.Common
make -C O.win32-x86 -f ../Makefile TOP=../../.. \
    T_A=win32-x86 install
make[1]: Entering directory 'J:/epics/devel-3.15/softGlue/softGlueApp/src/O.win32-x86'
perl -CSD H:/epics/base-3.15.4/bin/win32-x86/mkmf.pl  -m devAsynSoftGlue.d -I. -I../O.Common -I. -I. -I.. -I../../../inc
lude/compiler/msvc -I../../../include/os/WIN32 -I../../../include      -IJ:/epics/devel-3.15/asyn-4-30/include   -IJ:/ep
ics/devel-3.15/ipac-2-13/include    -IH:/epics/base-3.15.4/include/compiler/msvc -IH:/epics/base-3.15.4/include/os/WIN32
 -IH:/epics/base-3.15.4/include        devAsynSoftGlue.d ../devAsynSoftGlue.c
perl -CSD H:/epics/base-3.15.4/bin/win32-x86/mkmf.pl  -m drvIP_EP201.d -I. -I../O.Common -I. -I. -I.. -I../../../include
/compiler/msvc -I../../../include/os/WIN32 -I../../../include      -IJ:/epics/devel-3.15/asyn-4-30/include   -IJ:/epics/
devel-3.15/ipac-2-13/include    -IH:/epics/base-3.15.4/include/compiler/msvc -IH:/epics/base-3.15.4/include/os/WIN32 -IH
:/epics/base-3.15.4/include        drvIP_EP201.d ../drvIP_EP201.c
cl                -nologo -D__STDC__=0 -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE   -Ox -GL -Oy-   -W3 -w443
55        -MT -DEPICS_BUILD_DLL   -I. -I../O.Common -I. -I. -I.. -I../../../include/compiler/msvc -I../../../include/os/
WIN32 -I../../../include      -IJ:/epics/devel-3.15/asyn-4-30/include   -IJ:/epics/devel-3.15/ipac-2-13/include    -IH:/
epics/base-3.15.4/include/compiler/msvc -IH:/epics/base-3.15.4/include/os/WIN32 -IH:/epics/base-3.15.4/include        -c
 ../drvIP_EP201.c
drvIP_EP201.c
cl                -nologo -D__STDC__=0 -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE   -Ox -GL -Oy-   -W3 -w443
55        -MT -DEPICS_BUILD_DLL   -I. -I../O.Common -I. -I. -I.. -I../../../include/compiler/msvc -I../../../include/os/
WIN32 -I../../../include      -IJ:/epics/devel-3.15/asyn-4-30/include   -IJ:/epics/devel-3.15/ipac-2-13/include    -IH:/
epics/base-3.15.4/include/compiler/msvc -IH:/epics/base-3.15.4/include/os/WIN32 -IH:/epics/base-3.15.4/include        -c
 ../devAsynSoftGlue.c
devAsynSoftGlue.c
link -nologo /subsystem:windows /dll -LTCG -incremental:no -opt:ref -release       -out:softGlue.dll -implib:softGlue.li
b         drvIP_EP201.obj devAsynSoftGlue.obj      J:/epics/devel-3.15/asyn-4-30/lib/win32-x86/asyn.lib  J:/epics/devel-
3.15/ipac-2-13/lib/win32-x86/Ipac.lib  H:/epics/base-3.15.4/lib/win32-x86/dbRecStd.lib  H:/epics/base-3.15.4/lib/win32-x
86/dbCore.lib  H:/epics/base-3.15.4/lib/win32-x86/ca.lib  H:/epics/base-3.15.4/lib/win32-x86/Com.lib
   Creating library softGlue.lib and object softGlue.exp
drvIP_EP201.obj : error LNK2001: unresolved external symbol _pasynUInt32DigitalBase
drvIP_EP201.obj : error LNK2001: unresolved external symbol _pasynManager
drvIP_EP201.obj : error LNK2001: unresolved external symbol _pasynTrace
drvIP_EP201.obj : error LNK2001: unresolved external symbol _pasynInt32Base
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol _pasynEpicsUtils
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol _pdbbase
devAsynSoftGlue.obj : error LNK2001: unresolved external symbol _interruptAccept
softGlue.dll : fatal error LNK1120: 7 unresolved externals
H:/epics/base-3.15.4/configure/RULES_BUILD:281: recipe for target 'softGlue.dll' failed
make[1]: **\* [softGlue.dll] Error 1120
make[1]: Leaving directory 'J:/epics/devel-3.15/softGlue/softGlueApp/src/O.win32-x86'
H:/epics/base-3.15.4/configure/RULES_ARCHS:61: recipe for target 'install.win32-x86' failed
make: **\* [install.win32-x86] Error 2

**Comment by MarkRivers** [comment]
I just noticed something interesting.  All of the unresolved symbols are things that are declared "epicsShareExtern".  All of the epicsShareFunc symbols in asyn and base are being found OK.  I think this is probably an important observation.  With the epicsShareExtern symbols how does one module finally define that symbol as not extern?  Something seems to have changed with this between 3.14 and 3.15?

**Comment by MarkRivers** [comment]
This problem does not occur on mingw either statically or dynamically built.

**Comment by MarkRivers** [comment]
Problem fixed with commit #3 .
